### ISSUE TITLE
Connect pimatic to another pimatic !?

### ISSUE THREAD

**Issue by kenci** [pimatic/pimatic#434]
I have one RPi called "pool" and another one called "home". The "home"-pi is the main controller for switching lights, shutters and so on. Now i have this "pool"-pi outside in my garden and wants to switch the pool pump and pool lights with the "home"-pi. Can someone help me to achieve that?

"home-pi" -> sending -> -> receiving "pool"-pi -> switching pump...

**Comment by georg90** [comment]
I guess you won't need two pimatics connected. I have two Installation running, and what I really miss is something to connect to pimatic-homeduino from the other pimatic instance. (like pilight can)

So a possible Use Case could be:
You have a big home and one sender/receiver just isn't enough. You can place Pi's around the house with just sending/reciving all signals pimatic-homeduino sends to them.

**Comment by incmve** [comment]
I have two Pi's with pimatic that talk via the API, works like a charm.

**Comment by georg90** [comment]
can you explain?
Like I want to switch on a light and only one pimatic instance has a 433mhz sender.

**Comment by incmve** [comment]
I change variables but you could also switch on a device
http://pimatic.org/api/actions/#devices

I use this to change a variable, haven't tried a switch yet
```
IF $wind changes
THEN execute "curl --silent -X PATCH --header \"Content-Type:application/json\" --user \"USER:PASS\" --data '\{\"type\": \"value\", \"valueOrExpression\": $weather-station.avgAirspeed\}' http://192.168.0.12/api/variables/windspeedavg > /dev/null"
```

**Comment by leader21** [comment]
the problem at larger flats or houses is the range of the RF transmitter and receiver. when working with several pi's and homeduino you're having a bigger range and maybe (?) less cpu load due to less RF traffic at each.

well, i guess when having a "network" of pimatic pi's with homeduino or whatever pc you're using, it would be good to have all data visible at a "master frontend" then (and of course an own frontend for each pi too). so you can use all pi's in several rooms for sending and receiving only the information for the especially configured devices of this one pi that is "triggered" by action. if there would be an action on this pi, it will broadcast over wlan or ethernet to the others (or maybe only the "master pi") what actually changed. the master pi has the master frontend installed where all information will be collected together. some kind of "admin" frontend. 
guess the thing with the devices would not be that big problem (pilight did it, so pimatic can do it better ^^), but what about the rules and variables etc...

well, i'm no coder, just a dreamer ... maybe Oliver will implement this when he's bored sometime ^^

**Comment by georg90** [comment]
Yes, that's the way I use, too. :) thought of a more "native" implementation.
But thats nothing important for now.

**Comment by incmve** [comment]
it would be cool if pimatic had an master/slave function but the API work for me to.

**Comment by koffienl** [comment]
Can @sweetpi confirm if this feature is on the roadmap ?

**Comment by sweetpi** [comment]
I would love to have some kind peer to peer feature. So not one master config, but a separate config for each peer, so if one instance fails not all other instances are going down too.

Handling device of an remote peer isn't hard. My current idea is that you can add devices of another peer as some RemoteDevice to every peer instamce and all device  actions are invoked transparently over the netwok and attribute changes are propagated.

The rule system is a little bit more tricky. Should each rule be assigned to a specific peer? Should there be one master peer that handles all rules? I would prefer individual rules per peer. What do you think about handling rules?

**Comment by koffienl** [comment]
I would do something like Pilight : give each instance a unique ID.
Over the p2p network, you can identify devices as uid.devicename
For rules you could do the same : <code>IF uid1.devicename is on THEN turn uid2.devicename off</code>
You could make the UID name configurable in the config (also the option to be in the p2p).
When going to the add devices section, you would see 'local' devices and devices from other pimatic instances advertised over the network. You can then add them to the GUI.

During discussing this matter overhere, we came with several options:

- Make p2p available optional
- Make UID auto-generate and optional
- Perhaps a p2p groupname, sou you could make more than 1 networks?
- Make it possible to read or change devices from a rule on one pimatic to a other
- Make it possible to see rules on other pimatic instances in the p2p group

**Comment by sweetpi** [comment]
> I would do something like Pilight : give each instance a unique ID.

Yes, agree.

> Over the p2p network, you can identify devices as uid.devicename
> For rules you could do the same : IF uid1.devicename is on THEN turn uid2.devicename off

I wouldn't do that. Why not just force the user to define a unique id for every device in the network? If we get some id collision then I would ignore the remote device and emit a warning or error. Not prefixing with the id has the advantage that you can freely move the device from one instance to another without the need to adapt all references to this id. If someone wants to have ids prefixed with the instance uid he still can prefix ids manually.

> When going to the add devices section, you would see 'local' devices and devices from other pimatic instances advertised over the network. You can then add them to the GUI.

We need a device auto discover service (would for example replace debug outputs from homeduino and pimatic-max) anyway:
Discover Button => all plugins scan for devices => User gets a list of devices => User adapts config (if adaptable) and adds device to config.
This would then list all remote device, too. 

With the remote device approach we wouldn't have to define p2p-groups or so. You just add the device you need to the instance where you need it.

**Comment by georg90** [comment]
I like the approach @sweetpi - that's one of the strenghts of a pilight & pimatic Setup - you get the (awesome) pimatic and receving/sending is done via a p2p network of pilight instances.
That's why I thought of just extending pimatic-homeduino to be able to run by itself and "copy" all commands it get's from a master homeduino instance.

Question is how important is it to have this in overall pimatic? It looks like a lot of work to implement this instead of just using homeduino "like" pilight with different instances talking to each other.

**Comment by leader21** [comment]
:+1: i like it too :)
now you made me very curious ^^

**Comment by koffienl** [comment]
So with this approach you could make rules so it is defined on pi1 but switches a devices on pi2 ?

**Comment by georg90** [comment]
I like to pick up this topic.

@koffienl Yes, that's the intention I guess :-)
Is anything done in this direction?
What's the best approach? Implement a native pimatic solution or extend pimatic-homeduino?

I think of something which runs on an arduino, because:
- cheap
- less power consumption
- Pi with pimatic is just "overkill" if you just want to extend range/stability
- smaller, can be build in a box and be just plugged in an outlet (like a wlan extender)
- Have an arduino with a WLAN Interface (or LAN) --> Connects to pimatic
- TX/RX on arduino, logic on pimatic (same as now)

This is the pilight approach. So you would declare a device on one pimatic installation and it connects to the homeduino p2p network, any command is send from all homeduino nodes and received by them. 
In pimatic you can define which nodes you wanna use or if any. So you wont have a master/slave config, since you just use homeduino on the pimatic instances you wanna use it.

What are the steps needed to implement something like this? Any better approaches?

**Comment by bjornhij** [comment]
How about using xbee for communication between the arduinos? That way you wont have to configure a WLAN connection or have a wired LAN interface available.

**Comment by AvdH** [comment]
I agree with @georg90  (and @bjornhij). 
A seperate Arduino is the way to go, and using XBEE would be nice, because thn you could use another frequency (not flooding the 433Mhz) and opens the door to many more ready made sensiors.

**Comment by neographikal** [comment]
+1 for this as well, really would like to extend the range of my 433mhz radios. If it could be done using just an Arduino it would be great, less complicated, less expensive and lower power consumption

**Comment by sebster6** [comment]
I currently have two Pis in my home:
* One in the attic with a homebuilt double quad antenna (almost covers the whole house to switch 433 Mhz power switches), a Pi camera sticking out of the roof used for a time lapse of the building site behind our house (www.twitter.com/adenbuettel), and several inside/outside sensors (DHT22, DS18B20)
* Another one in the living room powering another DHT22 sensor and a 4x20 display on the wall, and also sending 433 Mhz signals to ensure correct switching when using Pimatic.

At the moment, I'm using Pilight as master and slave instances on the two Pis, and using Pimatic to bring everything together nicely.

So, my issue is on how to switch to Pimatic altogether without losing functionality. Since there is no master/slave at the moment, is there any chance to keep my current setup at all? Could an update to Pimatic break the Pilight functionality at any time? (I'm not updating Pilight to prevent that from happening, but I do update Pimatic on a regular basis.)
